aten/src/ATen/native/quantized/cpu/qconv_dynamic.cpp

```diff
@@ -0,0 +1,184 @@
+
+#ifdef USE_FBGEMM
+
+template <int kSpatialDim>
+at::Tensor PackedConvWeight<kSpatialDim>::apply_dynamic(
+    const at::Tensor& input,
+    bool reduce_range) {
+  /// is this needed?
+  auto input_contig = input.contiguous();
+  const auto* input_ptr = input_contig.data_ptr<float>();
+
+  TORCH_CHECK(
+      fbgemm::fbgemmSupportedCPU(), "Your CPU does not support FBGEMM.");
+
+  float x_min, x_max;
+  fbgemm::FindMinMax(
+      /*m=*/input_ptr,
+      /*min=*/&x_min,
+      /*max=*/&x_max,
+      /*len=*/input.numel());
+
+  // Input tensor is quantized as 8-bit unsigned values
+  static constexpr int precision = 8;
+  static constexpr bool is_signed = false;
+
+  // Calculate scale and zero point for quantization of input tensor
+  auto q_params = quant_utils::ChooseQuantizationParams(
+      /*min=*/x_min,
+      /*max=*/x_max,
+      /*qmin=*/is_signed ? -(1 << (precision - 1)) : 0,
+      /*qmax=*/
+      is_signed ? ((1 << (precision - 1)) - 1) : (1 << precision) - 1,
+      /*preserve_sparsity=*/false,
+      /*force_scale_power_of_two=*/false,
+      /*reduce_range=*/reduce_range);
+
+  // q_params.precision = precision;
+
+  // Quantize input
+  at::Tensor q_input = at::quantize_per_tensor(
+      input, q_params.scale, q_params.zero_point, c10::kQUInt8);
+
+  at::Tensor out = apply_impl<false>(input, q_params.scale, q_params.zero_point);
+  
+  return at::dequantize(out);
+}
+
+template at::Tensor PackedConvWeight<2>::apply_dynamic(
+    const at::Tensor& input,
+    bool reduce_range);
+
+template at::Tensor PackedConvWeight<3>::apply_dynamic(
+    const at::Tensor& input,
+    bool reduce_range);
+
+#endif // USE_FBGEMM
+
+#ifdef USE_PYTORCH_QNNPACK
+
+template <int kSpatialDim>
+at::Tensor PackedConvWeightsQnnp<kSpatialDim>::apply_dynamic(
+    const at::Tensor& input,
+    bool reduce_range) {
+
+    // On empty input, no output data will be generated,
+    // so use arbitrary qparams.
+    float x_min = 0;
+    float x_max = 0;
+    // Otherwise...
+    if (input.numel() > 0) {
+      x_min = input.min().item<float>();
+      x_max = input.max().item<float>();
+    }
+
+    // Input tensor is quantized as 8-bit unsigned values
+    static constexpr int precision = 8;
+    static constexpr bool is_signed = false;
+
+    // Calculate scale and zero point for quantization of input tensor
+    auto q_params = quant_utils::ChooseQuantizationParams(
+        /*min=*/x_min,
+        /*max=*/x_max,
+        /*qmin=*/is_signed ? -(1 << (precision - 1)) : 0,
+        /*qmax=*/
+        is_signed ? ((1 << (precision - 1)) - 1) : (1 << precision) - 1,
+        /*preserve_sparsity=*/false,
+        /*force_scale_power_of_two=*/false,
+        /*reduce_range=*/reduce_range);
+
+    // Quantize input
+    at::Tensor q_input = at::quantize_per_tensor(
+        input, q_params.scale, q_params.zero_point, c10::kQUInt8);
+
+    at::Tensor out = apply_impl<false>(q_input, q_params.scale, q_params.zero_point);
+
+    return at::dequantize(out);
```

Comment:
Put a comment here to the effect that we can have optimized kernel that produces fp32 output. This also holds for fbgemm.